### PR TITLE
Creation of the BatchnormAdaptationAlgorithm abstraction

### PR DESCRIPTION
**Pull request by andrey-churkin**

### CODE REVIEW

File: beta/nncf/tensorflow/batchnorm_adaptation.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    """
+
+    def run(self, model, config: NNCFConfig):
+        """
+        Runs the batch-norm adaptation algorithm.
+
+        :param model: A model for which the algorithm will be applied.
+        :param config: NNCF config.
+        """
+        # TODO(andrey-churkin): Should be implemented.
```
Comment (andrey-churkin) [inline beta/nncf/tensorflow/batchnorm_adaptation.py:30]: I created a ticket (55671) to track the progress of this task.

**Review by vshampor — COMMENTED** [review]
Looks good, but could you please, either in this PR or another, perform a more explicit specification for what the BN adaptation algorithm requires as parameters in order to function? Take a look at the `RangeInitParams`, `PrecisionInitParams`, how they are created and used and implement something like that for BN adapt.

**Comment by andrey-churkin** [comment]
> Looks good, but could you please, either in this PR or another, perform a more explicit specification for what the BN adaptation algorithm requires as parameters in order to function? Take a look at the `RangeInitParams`, `PrecisionInitParams`, how they are created and used and implement something like that for BN adapt.

I agree with this statement. It's a good idea. I suggest doing it on the ticket (55671) because it is necessary to introduce the BNAdaptationInitArgs extra config struct for NNCF TF. The purpose of this PR is to introduce a common API for BN adaptation to implement common compression algorithms.

**Review by vshampor — APPROVED** [review]

File: nncf/common/batchnorm_adaptation.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    accuracy drop even before model training.
+    """
+
+    def __init__(self):
+        """
+        Initializes the batch-norm adaptation algorithm.
+        """
+        self._impl = factory.create_bn_adaptation_algorithm_impl()  # type: BatchnormAdaptationAlgorithmImpl
+
+    def run(self, model: ModelType, config: NNCFConfig):
```
Comment (alexsu52) [inline nncf/common/batchnorm_adaptation.py:49]: @andrey-churkin What do you think about initilalizing BatchnormAdaptationAlgorithm instance by NNCFConfig as following `def __init__(self, config: NNCFConfig)`

Also what do you think about explicity specifying input parameters insted of using NNCFConfig? I mean `num_bn_adaptation_samples: int = 2000, num_bn_forget_samples: int = 1000, extra_config:NNCFExtraConfigStruct` insted of `config: NNCFConfig`.

File: nncf/compression_method_api.py [hunk 3]
```diff
@@ -136,39 +134,6 @@ def statistics(self, quickly_collected_only=False):
             stats.update(self._model.statistics(quickly_collected_only))
         return stats
 
-    def run_batchnorm_adaptation(self, config):
```
Comment (alexsu52) [inline nncf/compression_method_api.py:139]: @vanyalzr @AlexKoff88 Please take a look.

**Review by alexsu52 — CHANGES_REQUESTED** [review]

File: nncf/common/batchnorm_adaptation.py (hunk 2 above)
Comment (alexsu52) [reply nncf/common/batchnorm_adaptation.py:49]: @vshampor, @andrey-churkin Let's do it in this PR.

File: nncf/compression_method_api.py (hunk 3 above)
Comment (vanyalzr) [reply nncf/compression_method_api.py:139]: Looks good to me. Makes more sense to abstract BN adaptation as a separate algorithm rather than keep it in CompressionAlgorithm's API.

**Review by vanyalzr — APPROVED** [review]

**Comment by vanyalzr** [comment]
Maybe out of the scope of this PR, but we could consider extending the list of module types that BN adaptation acts on beyond `torch.nn.modules.batchnorm.BatchNorm2d`, e.g. 1d BNs, LayerNorms, etc. This would require extending the `_apply_to_batchnorms` method of `DataLoaderBNAdaptationRunner`.

**Comment by alexsu52** [comment]
> 
> 
> Maybe out of the scope of this PR, but we could consider extending the list of module types that BN adaptation acts on beyond `torch.nn.modules.batchnorm.BatchNorm2d`, e.g. 1d BNs, LayerNorms, etc. This would require extending the `_apply_to_batchnorms` method of `DataLoaderBNAdaptationRunner`.

Thanks for your comments. @negvet please keep it in mind.

File: nncf/common/batchnorm_adaptation.py (hunk 2 above)
Comment (andrey-churkin) [reply nncf/common/batchnorm_adaptation.py]: Done.

**Comment by andrey-churkin** [comment]
Jenkins please retry a build

File: nncf/config/structure.py [hunk 4]
```diff
@@ -11,7 +11,43 @@
...
+from nncf.common.initialization import NNCFDataLoader
+
+
 class NNCFExtraConfigStruct:
     @classmethod
     def get_id(cls) -> str:
         raise NotImplementedError
+
+
+class BNAdaptationInitArgs(NNCFExtraConfigStruct):
```
Comment (vshampor) [inline nncf/config/structure.py:25]: You moved this from torch to, basically, a common folder. Is this object general enough to cover all framework cases?

Comment (andrey-churkin) [reply nncf/config/structure.py:25]: I think so at this stage.

File: beta/nncf/tensorflow/batchnorm_adaptation.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ limitations under the License.
+"""
+
+from nncf.common.batchnorm_adaptation import BatchnormAdaptationAlgorithmImpl
+from nncf.common.utils.logger import logger as nncf_logger
+
+
+class TFBatchnormAdaptationAlgorithmImpl(BatchnormAdaptationAlgorithmImpl):
+    """
+    Implementation of the batch-norm adaptation algorithm for the TensorFlow backend.
```
Comment (alexsu52) [inline beta/nncf/tensorflow/batchnorm_adaptation.py:20]:
```suggestion
    Implementation of the batch-norm statistics adaptation algorithm for the TensorFlow backend.
```
The 'batch-norm statistics adaptation' is used in the documentation. Please, update here and below.

File: nncf/common/batchnorm_adaptation.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+        :param model: A model for which the algorithm will be applied.
+        """
+        if self._impl:
+            self._impl.run(model)
+        else:
+            nncf_logger.warning(
+                'Could not run batchnorm adaptation as the adaptation data loader is not provided as an extra struct. '
+                'Refer to `NNCFConfig.register_extra_structs` and the `BNAdaptationInitArgs` class.'
+            )
```
Comment (alexsu52) [inline nncf/common/batchnorm_adaptation.py:105]: It looks like there is a right place for this warning in the `__init__` method.

File: nncf/common/batchnorm_adaptation.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        if data_loader:
+            num_bn_adaptation_steps = np.ceil(num_bn_adaptation_samples / data_loader.batch_size)
+            num_bn_forget_steps = np.ceil(num_bn_forget_samples / data_loader.batch_size)
+
+            self._impl = factory.create_bn_adaptation_algorithm_impl(data_loader,
+                                                                     num_bn_adaptation_steps,
+                                                                     num_bn_forget_steps,
+                                                                     device)
+
+    def run(self, model: ModelType):
```
Comment (alexsu52) [inline nncf/common/batchnorm_adaptation.py:93]:
```suggestion
    def run(self, model: ModelType) -> None:
```

File: nncf/common/batchnorm_adaptation.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        """
+        Initializes the batch-norm adaptation algorithm implementation.
+        """
+        self._data_loader = data_loader
+        self._num_bn_adaptation_steps = num_bn_adaptation_steps
+        self._num_bn_forget_steps = num_bn_forget_steps
+        self._device = device
+
+    @abstractmethod
+    def run(self, model: ModelType):
```
Comment (alexsu52) [inline nncf/common/batchnorm_adaptation.py:45]:
```suggestion
    def run(self, model: ModelType) -> None:
```

File: nncf/common/batchnorm_adaptation.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            statistics of the original model (using large momentum value for rolling
+            mean updates). The actual number of samples will be a closest multiple of
+            the batch size.
+        :param device:
+        """
+        if num_bn_adaptation_samples < 0:
+            raise ValueError('Number of adaptation samples must be >= 0')
+
+        self._impl = None
+        if data_loader:
```
Comment (alexsu52) [inline nncf/common/batchnorm_adaptation.py:84]: I guess you should raise an exception because data_loader is a required argument.

File: nncf/common/batchnorm_adaptation.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    the batch-norm adaptation algorithm inherit.
+    """
+
+    def __init__(self,
+                 data_loader: NNCFDataLoader,
+                 num_bn_adaptation_steps: int,
+                 num_bn_forget_steps: int,
+                 device: Optional[str] = None):
+        """
+        Initializes the batch-norm adaptation algorithm implementation.
```
Comment (alexsu52) [inline nncf/common/batchnorm_adaptation.py:37]: Please add docstring for the arguments because they have another meaning than arguments of the `BatchnormAdaptationAlgorithm`.

File: nncf/common/factory.py [hunk 11]
```diff
@@ -15,23 +15,48 @@
...
     elif __nncf_backend__ == 'TensorFlow':
         exporter = TFExporter(model, input_names, output_names, model_args)
 
     return exporter
+
+
+def create_bn_adaptation_algorithm_impl(data_loader: NNCFDataLoader,
+                                        num_bn_adaptation_steps: int,
+                                        num_bn_forget_steps: int,
+                                        device: Optional[str] = None):
```
Comment (alexsu52) [inline nncf/common/factory.py:47]:
```suggestion
                                        device: Optional[str] = None) -> BatchnormAdaptationAlgorithmImpl:
```

**Comment by andrey-churkin** [comment]
@alexsu52 Thanks for your comments! All comments were resolved. Please verify.

**Review by alexsu52 — APPROVED** [review]

**Comment by andrey-churkin** [comment]
Jenkins please retry a build

**Comment by andrey-churkin** [comment]
@vshampor Could you please review the last changes? (Commits: 3f085e8, 1e294dc, 788a96a)

**Comment by negvet** [comment]
There are mutual imports

**nncf/common/factory.py** 
`from nncf.common.batchnorm_adaptation import BatchnormAdaptationAlgorithmImpl`

**nncf/common/batchnorm_adaptation.py** 
`import nncf.common.factory as factory`

File: nncf/common/batchnorm_adaptation.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+from abc import ABC, abstractmethod
+
+import numpy as np
+
+from nncf.api.compression import ModelType
+from nncf.common.initialization import NNCFDataLoader
+import nncf.common.factory as factory
+
+
+class BatchnormAdaptationAlgorithmImpl(ABC):
```
Comment (vshampor) [inline nncf/common/batchnorm_adaptation.py:24]: `impl` referce to the implementation, not the interface. Should probably call this `IBatchnormAdaptationRunner`.

File: nncf/common/batchnorm_adaptation.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        :param num_bn_forget_steps: Number of batches from the training dataset to pass
+            through the model at initialization in order to erase batch-norm statistics of
+            the original model.
+        :param device: Device to perform initialization. If `device` is `None` then the device
+            of the model parameters will be used.
+        """
+        self._data_loader = data_loader
+        self._num_bn_adaptation_steps = num_bn_adaptation_steps
+        self._num_bn_forget_steps = num_bn_forget_steps
+        self._device = device
```
Comment (vshampor) [inline nncf/common/batchnorm_adaptation.py:51]: Interfaces may not have an `__init__` method or any specified structure other than the method structure.

**Review by vshampor — CHANGES_REQUESTED** [review]

**Review by vshampor — APPROVED** [review]
